Improve spellcheck script to allow for projects to specify additional sources

adeljo-amd opened the pull request:
This PR contains the following update:
- Added config to spell-check C++ files in `.spellcheck.yaml`
- Added a file `.spellcheck.local.yaml` that can be used as a reference for other projects when they want to specify additional sources
- Added a python script `docs/yaml_merger.py` to merge any project's `.spellcheck.local.yaml` with the main `.spellcheck.yaml` in this repository. The script simply checks the nodes `name` and `source` and ensures they match before merging the additional sources (node depth must match too). Anything else is ignored.

neon60 commented on .github/workflows/linting.yml:
```diff
@@ -37,10 +37,26 @@ jobs:
+          python3 docs/yaml_merger.py .spellcheck.yaml .spellcheck.local.yaml .spellcheck.yaml
       - name: Run spellcheck
-        uses: rojopolis/spellcheck-github-actions@0.30.0
+        uses: rojopolis/spellcheck-github-actions@v0
```
Probably it's better to keep on fixed version, but we should update this. 

[0.46.0](https://github.com/rojopolis/spellcheck-github-actions/releases/tag/0.46.0)

adeljo-amd replied:
Fair enough, done

neon60 commented on .github/workflows/linting.yml:
```diff
@@ -37,10 +37,26 @@ jobs:
           curl --silent --show-error --fail --location https://raw.github.com/ROCm/rocm-docs-core/develop/.spellcheck.yaml -O
           curl --silent --show-error --fail --location https://raw.github.com/ROCm/rocm-docs-core/develop/.wordlist.txt >> .wordlist.txt
+      - name: Install python dependencies
+        run: pip install pyyaml
```
You can add this to requirements.in file. You will need to update requirements.txt file.

adeljo-amd replied:
It's already included actually (by default in all projects' requirements.txt), so I'll just remove that line

adeljo-amd replied:
Removed

neon60 commented on docs/yaml_merger.py:
This should not be in the docs root folder. Maybe should go to workflows?

https://github.com/ROCm/rocm-docs-core/tree/develop/.github/workflows

adeljo-amd replied:
Moved, seems like a fine place to put it, given only the workflow depends on it

neon60 commented on .spellcheck.yaml:
```diff
@@ -168,3 +167,13 @@ matrix:
   - pyspelling.filters.url:
+- name: Cpp
+  sources:
+  - ['*.cpp', '*.hpp']
```
By default, we should not check the cpp files. How can we make this happen?

adeljo-amd replied:
In spellcheck, none of the sources can be empty. `expect_match = false` only works when it can't find files that matches the pattern specified. I suppose the only way to disable this would be to use some conditionals, but because we're using `rojopolis/spellcheck-github-actions`, this would increase the complexity of the work just to find a workaround. 

The simplest way was to put the above, as all other projects ensure they place C++ files in sub-directories and not the parent directory

neon60 replied:
I suggest then to only check cpp and hpp files in docs folder by default.
```suggestion
  - ['docs/**/*.cpp', 'docs/**/*.hpp']
```

adeljo-amd replied:
That's a good compromise, added it in now

neon60 commented on .github/workflows/linting.yml:
```diff
@@ -37,10 +37,24 @@ jobs:
         if: ${{ ! contains( github.repository, 'rocm-docs-core') }}
         shell: sh
         run: |
+          curl --silent --show-error --fail --location https://raw.github.com/ROCm/rocm-docs-core/develop/docs/yaml_merger.py -O
```
Do you need to update the url after the yaml_merger.py move?

adeljo-amd replied:
Thanks for spotting that, I forgot about that line

adeljo-amd commented:
Here is a test action in the HIP repository that shows the spellcheck run https://github.com/ROCm/HIP/actions/runs/12868847369/job/35990222242

samjwu reviewed (commented):
I think this may belong to ci: instead of feat: since it's affecting workflow files (for the commitizen commit message)

is there a need to release a new version for this?

adeljo-amd commented:
> I think this may belong to ci: instead of feat: since it's affecting workflow files (for the commitizen commit message)
> 
> is there a need to release a new version for this?

Thanks for pointing that out, squashed and adjusted the commit message.

Spoke to @neon60 about the need to release and he says no.
